File: contest1/canPartitionKSubsets/canPartitionKSubsets.cpp
class Solution {
public:
    bool canPartitionKSubsets(vector<int>& nums, int k) {
        //a brute force solution
        int sum = 0;
        for(int i = 0; nums.size(); i++)sum += nums[i];
        if(k <= 0 || sum % k != 0)return false; //handle edge case
        vector<bool> v(nums.size(),false); //keep tracks with which number is used
        return search(nums,v,k,0,0,sum/k);
    }
    bool search(vector<int>& nums, vector<bool> v, int k, int index, int sum, int target){
        if(k == 1)return true; //1 group does not need partition
        if(sum == target)return search(nums,v,k-1,0,0,target); //start from the begining
        for(i = index; i < nums.size(); i++){
            if(!v[i] and nums[i] + sum <= target){// nums[i] not used and can be put in this group
                v[i] = true;// use it now 
                if(search(nums,v,k,i+1,sum + nums[i],target))return true;
                // store nums[i] in current group and search given that nums[i] is used 
                v[i] = false; // also search if nums[i] is not used
            }
        }
        return false;
    }
}

File: virtual contest/contest1/canPartitionKSubsets/canPartitionKSubsets.py
# true iff can partition nums[0..len(nums)] into K equal sum subsets
#first solution is brute force
class Solution(object):
    def canPartitionKSubsets(self, nums, k):
        target, rem = divmod(sum(nums), k)
        if rem: return False # k must divide sum(nums)
        #search function in function body
        #treat nums,target as global (in function scope)
        def search(groups):
            if not nums: return True
            v = nums.pop()
            for i, group in enumerate(groups): #iterate list, with indices
                if group + v <= target: #can allocate v into group[i]
                    groups[i] += v #update value of groups[i]
                    if search(groups): return True #if this state leads to a good end
                    groups[i] -= v #else, restore groups[i]
                if not group: break #the configuration of putting v into group[i] is a dead end
                # in another word, no matter where u put v in, it will lead to a dead end
                # thus we break this loop and return false
                # to save some time (as the following groups must be zeros)
            nums.append(v)
            return False
        
        nums.sort() #sort nums first
        if nums[-1] > target: return False #if the last (maximal) element in nums is greater than target
        while nums and nums[-1] == target: #while nums is not empty and the last element of nums is eq to target
            nums.pop() #pop it out
            k -= 1 #its a group of itself

        return search([0] * k) #do search on the remaining nums and groups

"""
        Time complexity: O(k!k^(N-k))

"""
# second solution applies dynamic programming
# by keeping track with memo
class Solution(object):
    def canPartitionKSubsets(self, nums, k):
        target, rem = divmod(sum(nums), k)
        if rem or max(nums) > target: return False
        visited = [False] * len(nums)
        
        def search(index,groups,curr_sum):
            print(index,groups,curr_sum)
            if groups == 1: 
                return True
            if curr_sum == target:
                return search(0,groups-1,0)
            i = index
            while i < len(nums):
                if not visited[i] and nums[i] + curr_sum <= target:
                    visited[i] = True
                    if search(i+1,groups,curr_sum+nums[i]):
                        return True
                    visited[i] = False
                i += 1
            return False
        
        return search(0,k,0)
        

#Time complexity: O(N*2^N)
